README.md
3gpi-utils
==========

3GPi を操作するためのツール類を提供します。

## 提供ファイル
次のファイルがパッケージに含まれています。

### /etc/default/3gpi-utils  
3GPi の動作設定を行うためのファイルです。  
設定可能な項目を次に示します。

+ CONSOLE_PORT  
  3GPi で AT コマンドを実行する際に使用するポートを設定します。  
  デフォルトは /dev/tty3GPI です。

+ RI_PIN  
  3GPi の着信通知に使用する GPIO のピン番号を指定します。  
  3GPi 基板上のジャンパピンで GPIO16 と GPIO18 を選択できます。  
  空白の場合は無効になります。  
  デフォルトは無効です。

+ POWER_PIN  
  3GPi の電源操作に使用する GPIO のピン番号を指定します。  
  3GPi 基板上のジャンパピンで GPIO17 と GPIO12 を選択できます。  
  空白の場合は無効になります。  
  デフォルトは 17 です。

+ STATUS_PIN  
  3GPi の電源状態取得に使用する GPIO のピン番号を指定します。  
  3GPi 基板上のジャンパピンで GPIO22 と GPIO13 を選択できます。  
  空白の場合は無効になります。  
  デフォルトは 22 です。

+ RESET_PIN  
  3GPi のリセット操作に使用する GPIO のピン番号を指定します。  
  3GPi 基板上のジャンパピンで GPIO27 と GPIO6 を選択できます。  
  空白の場合は無効になります。  
  デフォルトは 27 です。

+ TIMEOUT  
  3GPi の電源状態の検出時間を設定します。  
  単位は秒です。  
  デフォルトは 10 です。

+ AUTO_ON  
  Raspberry Pi ブート時の 3GPi の動作を設定します。  
  デフォルトは 1 です。

  1 : 3GPi の通信モジュールの電源をオンにします。  
  0 : 3GPi の操作は行いません。

+ AUTO_OFF  
  Raspberry Pi シャットダウン時の 3GPi の動作を設定します。  
  デフォルトは 1 です。

  1 : 3GPi の通信モジュールの電源をオフにします。  
  0 : 3GPi の操作は行いません。

+ AUTO_RESET  
  Raspberry Pi シャットダウン時の 3GPi の動作を設定します。  
  有効にするには AUTO_OFF=0 である必要があります。  
  デフォルトは 0 です。

  1 : 3GPi の通信モジュールの電源をリセットします。  
  0 : 3GPi の操作は行いません。

+ AUTO_GPS  
  Raspberry Pi 起動時の GPS の動作を設定します。  
  有効にするには AUTO_ON=1 である必要があります。  
  デフォルトは 0 です。

  1 : GPS を有効にします。  
  0 : GPS の操作は行いません。

+ WAKE_ON_RING  
  待機中に着信があった場合 RI 信号を出力します。  
  有効にするには AUTO_OFF=0 である必要があります。  
  デフォルトは 0 です。

+ STORE_SMS_ON_RING  
  待機中に SMS の着信があった場合に本文を保存します。  
  有効にするには WAKE_ON_RING=1 である必要があります。  
  デフォルトは 0 です。  
  **SORACOM グローバル向け Air SIM を使用する場合は 0 に設定してください。**

### /etc/modules-load.d/3gpi.conf  
3GPi に必要なカーネルモジュールを記述したファイルです。

### /usr/sbin/3gpictl  
3GPi の操作を行うための実行ファイルです。  
使用可能なオプションを次に示します。

--poweron : 電源をオンします。  
--poweroff : 電源をオフします。  
--reset : 電源をリセットします。  
--status : 電源状態を表示します。  
--riclear : RI をクリアします。  
--rienable : RI を有効にします。  
--gpson : GPS を有効にします。  
--gpsoff : GPS を無効にします。  
--version, -v : バージョンを表示します。  
--help, -h : ヘルプを表示します。

### /usr/lib/3gpi-utils/gpio-export.sh  
3GPi で使用する GPIO ポートを初期化する実行ファイルです。

### /usr/lib/3gpi-utils/gpio-unexport.sh  
3GPi で使用する GPIO ポートを開放する実行ファイルです。

### /usr/lib/3gpi-utils/modem-init.sh  
3GPi のモデム初期化を行う実行ファイルです。

### /usr/lib/3gpi-utils/start.sh  
3GPi の起動処理を行う実行ファイルです。

### /usr/lib/3gpi-utils/stop.sh  
3GPi の終了処理を行う実行ファイルです。

### /usr/lib/3gpi-utils/chatscripts/gps-off  
GPS の動作を無効にする AT コマンドを発行します。

### /usr/lib/3gpi-utils/chatscripts/gps-on  
GPS の動作を有効にする AT コマンドを発行します。

### /usr/lib/3gpi-utils/chatscripts/ri-clear  
RI 信号をクリアする AT コマンドを発行します。

### /usr/lib/3gpi-utils/chatscripts/ri-enable-me  
待機中の RI 信号の出力を有効にする AT コマンドを発行します。  
着信時に受信した SMS の本文は保存します。

### /usr/lib/3gpi-utils/chatscripts/ri-enable-sm  
待機中の RI 信号の出力を有効にする AT コマンドを発行します。  
着信時に受信した SMS の本文は廃棄します。

### /lib/udev/rules.d/70-3gpi-usbserial.rules  
3GPi のデバイスを定義した設定ファイルです。

### /lib/systemd/system/3gpi-modem-init.service  
3GPi のモデム初期化サービスの設定ファイルです。

### /lib/systemd/system/3gpi-utils.service  
3GPi のサービスの設定ファイルです。

### /usr/share/bash-completion/completions/3gpictl
3gpictl のコマンド補完を行うための設定ファイルです。

### /usr/share/doc/3gpi-utils/changelog.gz
パッケージの変更履歴を記録したファイルです。

### /usr/share/doc/3gpi-utils/copyright
著作権とライセンスを記載したファイルです。


debian/changelog
3gpi-utils (2.0.2) buster; urgency=medium

  * bash_completion.d/3gpictl: Fix option completion
  * Update debhelper compatibility level to 12
  * Bump Standards-Version to 4.3.0

 -- Masahiro Honda <honda@mechatrax.com>  Mon, 08 Jul 2019 16:11:05 +0900

3gpi-utils (2.0.1) unstable; urgency=medium

  * Fix dependencies

 -- Masahiro Honda <honda@mechatrax.com>  Mon, 25 Dec 2017 17:57:46 +0900

3gpi-utils (2.0.0) unstable; urgency=low

  * debian/*: Change to provide multiple binary packages
  * Add modem-init.sh
  * Add 3gpi-utils-modem-init.service
  * Add chatscripts
  * Add bash_completion for 3gpictl
  * Update package maintainer scripts

 -- Masahiro Honda <honda@mechatrax.com>  Fri, 15 Dec 2017 17:43:02 +0900

3gpi-utils (1.0.3) unstable; urgency=low

  * debian/changelog: Fix typo
  * debian/control: Change Homepage URL
  * debian/copyright: Change Format URL
  * Remove invoking NetworkManager from package maintainer scripts
  * Change package priority from extra to optional
  * Update Standards-Version to 3.9.8

 -- Masahiro Honda <honda@mechatrax.com>  Fri, 15 Sep 2017 10:26:58 +0900

3gpi-utils (1.0.2) unstable; urgency=low

  * Fix URL typo

 -- Masahiro Honda <honda@mechatrax.com>  Tue, 22 Aug 2017 18:03:35 +0900

3gpi-utils (1.0.1) unstable; urgency=medium

  * Fix configuration description

 -- Masahiro Honda <honda@mechatrax.com>  Mon, 08 May 2017 15:29:09 +0900


3gpi-utils (1.0.0) unstable; urgency=medium

  * Switch to Native Debian Package
  * Add 3gpi.conf
  * Add export.sh
  * Add start.sh
  * Add stop.sh
  * Add unexport.sh
  * Change 3gpi-utils.service to execute scripts in /usr/lib/3gpi-utils
  * Change 3gpictl to read settings from /etc/default/3gpi-utils

 -- Masahiro Honda <honda@mechatrax.com>  Wed, 07 Dec 2016 19:28:10 +0900

3gpi-utils (0.5-1) unstable; urgency=medium

  * Update 3gpictl

 -- Masahiro Honda <honda@mechatrax.com>  Mon, 17 Oct 2016 11:15:22 +0000

3gpi-utils (0.4-1) unstable; urgency=medium

  * Update 3gpictl
  * Fix 3gpi-utils.service
  * Fix copyright description
  * Fix package description

 -- Masahiro Honda <honda@mechatrax.com>  Fri, 04 Dec 2015 01:33:05 +0000

3gpi-utils (0.3-1) unstable; urgency=medium

  * Fix 3gpi-utils.service
  * Modify 3gpictl
  * Modify package maintainer scripts
  * Modify udev rule

 -- Masahiro Honda <honda@mechatrax.com>  Tue, 13 Jan 2015 12:18:51 +0900

3gpi-utils (0.2-1) unstable; urgency=low

  * Add copyright
  * Use dh-systemd
  * Modify scripts

 -- Masahiro Honda <honda@mechatrax.com>  Mon, 17 Nov 2014 12:53:06 +0000

3gpi-utils (0.1-1) unstable; urgency=low

  * Initial release

 -- Masahiro Honda <honda@mechatrax.com>  Thu, 11 Sep 2014 06:08:08 +0000


3gpiinit.sh
#!/bin/sh

#
# initialize 3gpi module script
# 

POWER_PIN="17"
POWER_DIR="out"

RESET_PIN="27"
RESET_DIR="out"

STATUS_PIN="22"
STATUS_DIR="in"

EXPORT="/sys/class/gpio/export"

SETTING="/etc/default/3gpi-utils"

#
# Usage: export_gpio [GPIO_PIN]
#
export_gpio ()
{
    DIRECTION="/sys/class/gpio/gpio${1}/direction"

    if [ ! -e $DIRECTION ]
    then
        echo $1 > $EXPORT
    fi
}

#
# Usage: set_direction [GPIO_PIN] [DIRECTION]
#
set_direction ()
{
    DIRECTION="/sys/class/gpio/gpio${1}/direction"

    if [ ! -w $DIRECTION ]
    then
        echo "cannot access ${DIRECTION}"
        return 1
    fi
    
    echo $2 > $DIRECTION
}

#
# initialize 3gpi
#

if [ -e $SETTING ]
then
    . $SETTING
else
    exit 1
fi

if [ ! -w $EXPORT ]
then
    echo "cannot access ${EXPORT}"
    exit 1
fi

if ! (export_gpio $POWER_PIN &&
    export_gpio $RESET_PIN &&
    export_gpio $STATUS_PIN)
then
    echo "gpio export failure"
    exit 1
fi

sleep 1

if ! (set_direction $POWER_PIN $POWER_DIR &&
    set_direction $RESET_PIN $RESET_DIR &&
    set_direction $STATUS_PIN $STATUS_DIR)
then
    exit 1
fi

if [ $AUTO_ON -eq 1 ]
then
    if [ $(/usr/sbin/3gpictl --status) = "off" ]
    then
        /usr/sbin/3gpictl --poweron
    fi
    if [ $AUTO_GPS -eq 1 ]
    then
        /usr/sbin/3gpictl --gpson
    fi
fi

